mimxrt: Add the hw i2c class to machine.

robert-hh opened the pull request:
Advantage over SoftI2C:
- Full protocol support.
- Higher data rate -  up to ~3 MHZ.

robert-hh committed:
mimxrt: Add HW machine.i2c to the module machine.

It uses non-blocking transfer of data. Advantage over SoftI2C:
- Higher data rate up to 1.54 MHZ.
- Full protocol support.

robert-hh committed:
mimxrt: Force stop mode if requestet.

Even in case of an error.

robert-hh committed:
mimxrt: Add the I2C defines for the other boards.

MIMXRT1010_EVK
MIMXRT1050_EVK
MIMXRT1060_EVK
MIMXRT1064_EVK
Teensy 4.0
Teensy 4.1

robert-hh committed:
mimxrt: Further extensions and small changes.

- Attempt to support 10 bit addresses.
- Extend HW I2C clock range to ~3 MHz.
- Use Pin pull-up instead of Pin hold, preventing a lock in
  case of open wires.

kadamski commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -50,10 +50,9 @@ typedef struct _machine_i2c_obj_t {
-    bool nack_flag;
     uint8_t mode;
+    bool transfer_busy;
+    status_t transfer_status;
```
Many of the changes in this commit are not related to the force stop mode and it would be better to just move them to the previous commit, which would make less confusion.

robert-hh replied:
This part of the code is changed in further commits. So I leave it as it is. And I will definitely not start to split up commits, which will be squashed at the end anyhow. Please make you comment to the final state of the files, not some intermediate states. That is pointless.

kadamski replied:
But how could I know if you intend to squash commits and which of them you intend to squash?

robert-hh replied:
Damien will probably squash all commits of this PR into a single one, when he merges the commit. So do not look at individual commits here, because the only purpose of the PR is providing HW I2C.

kadamski replied:
Ach, ok. I find this strategy strange as, well, what should be the commit message of such a squashed commit? :) But, well, its up to Damien. I'll be reviewing the result only from now on.

robert-hh commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,210 @@
+    }
+}
+
+    lpi2c_master_handle_t g_master_handle;
```
That line 165 does not exist in my file.

robert-hh commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,210 @@
+    machine_i2c_obj_t *self = m_new_obj(machine_i2c_obj_t);
+    self->base.type = &machine_hw_i2c_type;
+    self->i2c_id = i2c_id;
+    self->i2c_hw_id = i2c_index_table[i2c_id];  // the hw spi number 1..n
```
Indeed a copy/paste glitch from the template file in the comment.

robert-hh commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,210 @@
+    if (status == kStatus_LPI2C_Nak) {
+        self->nack_flag = true;
+    } else {
+        self->transfer_error = (status != kStatus_Success);
```
You are working on an old version of the file. So some of your comments do not apply

robert-hh commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,210 @@
+    self->i2c_id = i2c_id;
+    self->i2c_hw_id = i2c_index_table[i2c_id];  // the hw spi number 1..n
+    self->i2c_inst = i2c_base_ptr_table[self->i2c_hw_id];
+    self->mode = 0;
```
self->mode is not used. Was planned for master/slave mode.

robert-hh commented on ports/mimxrt/boards/MIMXRT1020_EVK/mpconfigboard.h:
```diff
@@ -13,7 +13,8 @@
 // Define mapping logical UART # to hardware UART #
-// D3/D5   LPUART1  Not usable, Since D3 is blocked.
+// RX/TX   HW-UART    Logical I2C
+// D3/D5   LPUART1    Not usable, Since D3 is blocked.
```
No. It is related to UART. But it is a part of the mpconfigport.h file.

robert-hh committed:
mimxrt: Remove the element mode from machine_i2c_obj_t.

It was foreseen for master/slave mode, which is not (yet) implemented.

Minor changes: Change a comment.

robert-hh commented:
@dpgeorge This PR is ready to go.

dpgeorge commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,217 @@
+
+STATIC const mp_machine_i2c_p_t machine_i2c_p = {
+    .transfer = mp_machine_i2c_transfer_adaptor,
+    .transfer_single = machine_i2c_transfer_single,
```
I guess you needed to use the adaptor because there was no easy way to transfer multiple segments?  It's not a problem, just wondering.

robert-hh replied:
It was just a copy of method wihtout much thinking.

robert-hh replied:
But in fact the NXP API performs all the frame handling like handling start and stop conditions,  and expect a single message to be sent.

dpgeorge commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,217 @@
+        MICROPY_EVENT_POLL_HOOK
+    }
+
+    // Transfer will not send a stop in case of errors like NAK. So it#s done here.
```
`it's` instead of `it#s`

robert-hh replied:
OK. Will be changed

dpgeorge commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,217 @@
+    .transfer_single = machine_i2c_transfer_single,
+};
+
+const mp_obj_type_t machine_hw_i2c_type = {
```
maybe call it `machine_i2c_type` to match `machine_spi_type`?

robert-hh replied:
I can change that without conflicts.

dpgeorge commented on ports/mimxrt/machine_i2c.c:
```diff
@@ -0,0 +1,217 @@
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_freq, MP_ARG_INT, {.u_int = DEFAULT_I2C_FREQ} },
+        { MP_QSTR_drive, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_I2C_DRIVE} },
```
what arguments can be passed to `drive`?

robert-hh replied:
The same as for the Pin module and and the SPI module. Values 0 - 6 or Pin.POWER_0 to Pin.POWER_6. It changes the hardware pin impedance.  Since pins are assigned fixed to the I2C id's, not Pin settings can be made. But obviously this is only critical for the low edge. And since the I2C speed is low, it is less important than for SPI.

robert-hh committed:
mimxrt: Change to machine_i2c.c as reviewed.

robert-hh commented:
The PR is updated with the suggested changes.

dpgeorge commented:
Thank you!  Squashed and merged in 98c5703027e7db9a030f747f4dd1684407230656

robert-hh commented:
Thanks a lot. The next PR is most likely Ethernet support for the MIMXRT boards and Teensy 4.1. As far as I can tell it is already working fine. I just keep it on the bench for a while for testing and further thoughts.
